File: LRUCache.py
from typing import List, Dict, Tuple

class LRUCache:

    def __init__(self, capacity: int):
        self.n = capacity
        self.dll = []
        self.mp = {}

    def makeRecentlyUsed(self, key: int) -> None:
        self.dll.remove(self.mp[key][0])
        self.dll.insert(0, key)
        self.mp[key] = (self.dll.index(key), self.mp[key][1])

    def get(self, key: int) -> int:
        if key not in self.mp:
            return -1
        self.makeRecentlyUsed(key)
        return self.mp[key][1]

    def put(self, key: int, value: int) -> None:
        if key in self.mp:
            self.mp[key] = (self.mp[key][0], value)
            self.makeRecentlyUsed(key)
        else:
            self.dll.insert(0, key)
            self.mp[key] = (self.dll.index(key), value)
            self.n -= 1
        if self.n < 0:
            keyToBeDeleted = self.dll.pop()
            del self.mp[keyToBeDeleted]
            self.n += 1

File: LRUCache.java
import java.util.*;

class LRUCache {

    private int n;
    private List<Integer> dll;
    private Map<Integer, Pair<List<Integer>.Iterator, Integer>> mp;

    public LRUCache(int capacity) {
        this.n = capacity;
        this.dll = new LinkedList<>();
        this.mp = new HashMap<>();
    }

    private void makeRecentlyUsed(int key){
        dll.remove(mp.get(key).getKey());
        dll.add(0, key);
        mp.put(key, new Pair<>(dll.listIterator(), mp.get(key).getValue()));
    }

    public int get(int key) {
        if(!mp.containsKey(key)) {
            return -1;
        }
        makeRecentlyUsed(key);
        return mp.get(key).getValue();
    }

    public void put(int key, int value) {
        if(mp.containsKey(key)) {
            mp.get(key).setValue(value);
            makeRecentlyUsed(key);
        } else {
            dll.add(0, key);
            mp.put(key, new Pair<>(dll.listIterator(), value));
            n--;
        }
        if(n < 0) {
            int keyToBeDeleted = dll.remove(dll.size() - 1);
            mp.remove(keyToBeDeleted);
            n++;
        }
    }
}

